src/lib/Components/Bidding/Components/__tests__/Timer-tests.tsx

```diff
@@ -94,3 +96,167 @@ it("shows month, date, and hour adjusted for the timezone where the user is", ()
+    })
+  })
+
+  describe("nextState", () => {
+    it("gives the correct next state for a sale in preview that is timed", () => {
+      expect(timer.nextState(AuctionTimerState.PREVIEW, null)).toEqual(AuctionTimerState.CLOSING)
+    })
+
+    it("gives the correct next state for a sale in preview that will go live", () => {
+      expect(timer.nextState(AuctionTimerState.PREVIEW, futureTime)).toEqual(
+        AuctionTimerState.LIVE_INTEGRATION_UPCOMING
+      )
+    })
+
+    it("gives the correct next state for a sale that has closed", () => {
+      expect(timer.nextState(AuctionTimerState.CLOSED, null)).toEqual(AuctionTimerState.CLOSED)
+    })
+
+    it("gives the correct next state for a sale that is about to close", () => {
+      expect(timer.nextState(AuctionTimerState.CLOSING, null)).toEqual(AuctionTimerState.CLOSED)
+    })
+
+    it("gives the correct next state for a sale that is about to go live", () => {
+      expect(timer.nextState(AuctionTimerState.LIVE_INTEGRATION_UPCOMING, futureTime)).toEqual(
+        AuctionTimerState.LIVE_INTEGRATION_ONGOING
+      )
+    })
+
+    it("gives the correct next state for a sale that is currently live", () => {
+      expect(timer.nextState(AuctionTimerState.LIVE_INTEGRATION_ONGOING, pastTime)).toEqual(
+        AuctionTimerState.LIVE_INTEGRATION_ONGOING
+      )
+    })
+  })
+
+  describe("upcomingLabel", () => {
+    it("shows the correct label for a sale in preview", () => {
+      const { label, relevantDate } = timer.upcomingLabel(AuctionTimerState.PREVIEW, null, futureTime, null)
+      expect(label).toEqual("Starts May 10, 5pm")
+      expect(relevantDate).toEqual(futureTime)
+    })
+
+    it("shows the correct label for a sale that is going to end", () => {
+      const { label, relevantDate } = timer.upcomingLabel(AuctionTimerState.CLOSING, null, pastTime, futureTime)
+      expect(label).toEqual("Ends May 10, 5pm")
+      expect(relevantDate).toEqual(futureTime)
+    })
+
+    it("shows the correct label for a sale that is going to go live", () => {
+      const { label, relevantDate } = timer.upcomingLabel(
+        AuctionTimerState.LIVE_INTEGRATION_UPCOMING,
+        futureTime,
+        pastTime,
+        null
+      )
+      expect(label).toEqual("Live May 10, 5pm")
+      expect(relevantDate).toEqual(futureTime)
+    })
+
+    it("shows the correct label for a sale that is already live", () => {
+      const { label, relevantDate } = timer.upcomingLabel(
+        AuctionTimerState.LIVE_INTEGRATION_ONGOING,
+        pastTime,
+        pastTime,
+        null
+      )
+      expect(label).toEqual("In progress")
+      expect(relevantDate).toEqual(null)
+    })
+
+    it("shows the correct label for a sale that has closed", () => {
+      const { label, relevantDate } = timer.upcomingLabel(AuctionTimerState.CLOSED, pastTime, pastTime, pastTime)
+      expect(label).toEqual("Bidding closed")
+      expect(relevantDate).toEqual(null)
+    })
+  })
```

Comment:
These new tests are all testing internals and it would make it more difficult to refactor this component. It is also easier to understand if it tests the behaviour of a component. Would it be possible to write a test that doesn't depend on `state` or internal functions? I usually use `state` as the last resort and add a comment for why calling `state` is absolutely necessary.

Reply:
Since these methods have a lot of logic and depend on each other, breaking them out to test them individually in this way was super helpful while developing! But let me take a stab at trying to create tests that don't rely on the exact internals. 👍